### ISSUE TITLE
Volumes V2 Should Expose Extensions

### ISSUE THREAD

**Issue by spjmurray** [gophercloud/gophercloud#184]
Volume lost the ability to parse `os-vol-tenant-attr:tenant_id` in the past.  It'd be nice to have access to this again for metric collection (e.g. we can map volume usage back to specific projects/domains and watch trends).  I'm willing to do the work just need some pointers on the best way to implement this extension.

**Comment by jrperritt** [comment]
Ah, right. Because there can be arbitrarily many extension fields, I thought it would be best to let users add them as-needed. In this case, it really just amounts to a new package for that extension with a `results.go` file with something like:
```go
package tenantattr

type VolumeExt struct {
  TenantID `json:"os-vol-tenant-attr:tenant_id"`
}
```

Then in your application code you'd have something like (untested):
```go
type MyVolume struct {
  volumes.Volume
  tenantattr.VolumeExt
}

var v struct {
  MyVolume `json:"volume"`
}

err := volumes.Get(client, volID).ExtractInto(&v)
fmt.Println(v.TenantID)
```

I should mention this is not in stone; I'm open to suggestions for a better way of implementing extensions [to response objects].

**Comment by spjmurray** [comment]
Got you. Perhaps a nice API to allow extraction of arbitrary extensions is a decent measure. Zero changes to library code and access to all the things from the client side

**Comment by jrperritt** [comment]
It's hard to say without seeing the coded implementation, but if you'd like to make a PR or provide your idea for the implementation here, maybe we can flush out the details.

**Comment by spjmurray** [comment]
Inbound tomorrow

**Comment by spjmurray** [comment]
In the end I just bodged it in client code https://github.com/spjmurray/telegraf/blob/openstack_stats/plugins/inputs/openstack/openstack.go#L246.  Is this the right/clean way to do it?  I guess if so this issue can be closed

**Comment by jtopjian** [comment]
Since this is an [official extension](https://github.com/openstack/cinder/blob/a15459a27777e98599d70392191b0a6f0288ff34/cinder/api/contrib/volume_tenant_attribute.py), I'd like to see this included and I could do the work if your current solution is working.

This extension feels awkward, though. Almost all other resources have a tenant or owner attribute, so IMO it's a little cumbersome to remember that Volumes are treated differently. I was thinking that maybe support for volume+tenant can be handled similarly to the [portsbinding](https://github.com/gophercloud/gophercloud/tree/master/openstack/networking/v2/extensions/portsbinding) extension (where there would be a full Get call in the tenantattr extension that would include all base attributes plus Tenant ID. This would probably work short-term, but I don't think this would scale well with further extended / custom attributes (if those ever existed).

**Comment by jrperritt** [comment]
> but I don't think this would scale well with further extended / custom attributes

For me, that's why that's undesirable. I'm glad you mentioned the `portsbinding` implementation, so that if we come up with a good way to handle extensions we can change that one as well.

The solution I posted above works, but is less than ideal. In addition, what @spjmurray did works as well, though every user would have to implement more of the code than if the extensions were available in Gophercloud. I will test a few different ideas to try to find a good way to handle this. Others are welcome to do the same.

**Comment by spjmurray** [comment]
The root of the problem IMO is that cinder should stop classifying this as an extension and make it part of the standard as Joe alluded to. Pain and suffering of making it happen aside it's worth a try making it happen. Or am I being naive? :smile:

**Comment by jrperritt** [comment]
One option is two-stepping the JSON unmarshalling (see #186). Thoughts on that @spjmurray @jtopjian ?

**Comment by spjmurray** [comment]
I'm more of a C++ Yoda than Go, so excuse my stupidity. Does this play well with a 'list detail' operation, and how does it fit in with other v2 extensions like migration and replication?

**Comment by jrperritt** [comment]
> Does this play well with a 'list detail' operation

The original commit doesn't address it, but the most recent one does.

> how does it fit in with other v2 extensions like migration and replication

They would be added in exactly the same way as `TenantID` was added.

**Comment by jtopjian** [comment]
@jrperritt #186 is pretty clever. If I understand it correctly, it allows the dev/user to compose a struct with the fields they'd like to see in the result, and `ExtractInto` will mash the composition together and present the result.

If that's correct, this method has the benefit of being really easy to add custom/vendor fields or if someone wanted to quickly support a field that is not yet in Gophercloud without risking breakage if/when the field was added.

I like it. :smile: 

Coincidentally enough, someone just requested the `OS-EXT-AZ:availability_zone` field of `servers` in Terraform. Timing couldn't be better.

**Comment by jrperritt** [comment]
>  this method has the benefit of being really easy to add custom/vendor fields or if someone wanted to quickly support a field that is not yet in Gophercloud without risking breakage if/when the field was added.

correct

**Comment by jrperritt** [comment]
closed by #186
